internal/api/auth_hooks.go

```diff
@@ -0,0 +1,164 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"net/url"
+
+	"fmt"
+	"github.com/gofrs/uuid"
+	"github.com/supabase/gotrue/internal/conf"
+	"github.com/supabase/gotrue/internal/storage"
+	"regexp"
+	"strings"
+)
+
+type HookType string
+
+const (
+	PostgresHook HookType = "pg-functions"
+	HTTPHook     HookType = "http"
+)
+
+const (
+	MFAHookRejection = "reject"
+	MFAHookContinue  = "continue"
+)
+
+type AuthHook struct {
+	ExtensibilityPointConfiguration conf.ExtensibilityPointConfiguration
+	payload                         []byte
+	hookType                        HookType
+	event                           string
+	db                              *storage.Connection
+}
+
+// Hook Events
+const (
+	MFAVerificationEvent = "auth.mfa_verfication"
+)
+
+type HookErrorResponse struct {
+	ErrorMessage string `json:"error_message"`
+	ErrorCode    string `json:"error_code"`
+	RetryAfter   bool   `json:"retry_after"`
+}
+
+func parseErrorResponse(response []byte) (*HookErrorResponse, error) {
+	var errResp HookErrorResponse
+	err := json.Unmarshal(response, &errResp)
+	if err != nil {
+		return nil, err
+	}
+	if errResp.ErrorMessage != "" {
+		return &errResp, nil
+	}
+	return nil, err
+}
+
+func parseMFAVerificationResponse(response []byte) (*MFAVerificationHookOutput, error) {
+	var MFAVerificationResponse MFAVerificationHookOutput
+	err := json.Unmarshal(response, &MFAVerificationResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MFAVerificationResponse, err
+}
+
+type MFAVerificationHookInput struct {
+	UserID   uuid.UUID `json:"user_id"`
+	FactorID uuid.UUID `json:"factor_id"`
+	Valid    bool      `json:"valid"`
+}
+type MFAVerificationHookOutput struct {
+	Decision string `json:"decision"`
+}
+
+// Functions for encoding and decoding payload
+func CreateMFAVerificationHookInput(user_id uuid.UUID, factor_id uuid.UUID, valid bool) ([]byte, error) {
+	payload := MFAVerificationHookInput{
+		UserID:   user_id,
+		FactorID: factor_id,
+		Valid:    valid,
+	}
+	data, err := json.Marshal(&payload)
+	if err != nil {
+		panic(err)
+	}
+	return data, nil
+}
+
+func (ah *AuthHook) Trigger() ([]byte, error) {
+	// Parse URI object
+	url, err := url.Parse(ah.ExtensibilityPointConfiguration.URI)
+	if err != nil {
+		return nil, err
+	}
+	// trigger appropriate type of hook
+	switch url.Scheme {
+	case string(PostgresHook):
+		return ah.triggerPostgresHook()
+	default:
+		return nil, errors.New("unsupported hook type")
+	}
+}
```

Comment:
I'm not super certain why we need this abstraction. Can't we just have a single function under `API` that does this very simply?

Example:

```go
func (a *API) invokeHook(ctx context.Context, input any) (any, error) {
  switch input.(type) {
    case *MFAVerificationInput:
      // trigger the MFA verification hook
     
  }
}
```

I want to avoid overly complex and unnecessary abstractions at this point, as I don't see what benefit comes out of them.

Reply:
You can also do something like to keep it typesafe:

```go
input := MFAVerificationAttemptInput{...}
output := MFAVerificationAttemptOutput{}

if err := a.invokeHook(ctx, input, output); err != nil {
  ...
}

if output.IsError() {
  // return error json
} else {
  // process the logic
}
```